**Problem 1:**

**Raghav had just sold his startup and suddenly found himself with a lot of downtime. He decided to binge-watch a TV series, something he had never done in his life. He didn’t ‘trust the ratings’ because he thought they were all paid for. He decided to start off by watching the TV series with the most number of ratings than the one with the highest rating score. Help Raghav in his quest to find that show using the datasets provided here - > https://datasets.imdbws.com/ Metadata info about those datasets are available here -> https://www.imdb.com/interfaces/. Implement the whole solution in code so that it is reproducible. Print the name of the TV series that received the most number of votes along with the number of votes it received.**

Importing pandas and numpy.

In [ ]:
import pandas as pd
import numpy as np

Mounting the drive and importing the required datasets - Title.Basics and Title.Ratings files. 

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')
root_path = 'gdrive/My Drive/BedRockAI/'  

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [ ]:
basics_tsv_file = root_path+ "title.basics.tsv.gz"
basics = pd.read_csv(basics_tsv_file, sep='\t', low_memory=False)
ratings_tsv_file = root_path+ "title.ratings.tsv.gz"
ratings = pd.read_csv(ratings_tsv_file, sep='\t',low_memory=False)

Displaying shape and sample data for Title and Ratings:

In [ ]:
basics.shape[0]

8694425

In [ ]:
basics.head()

tconst titleType  ... runtimeMinutes                    genres
0  tt0000001     short  ...              1         Documentary,Short
1  tt0000002     short  ...              5           Animation,Short
2  tt0000003     short  ...              4  Animation,Comedy,Romance
3  tt0000004     short  ...             12           Animation,Short
4  tt0000005     short  ...              1              Comedy,Short

[5 rows x 9 columns]

In [ ]:
ratings.shape[0]

1214631

In [ ]:
ratings.head()

tconst  averageRating  numVotes
0  tt0000001            5.7      1858
1  tt0000002            6.0       243
2  tt0000003            6.5      1630
3  tt0000004            6.0       158
4  tt0000005            6.2      2456

Combining Title and Ratings based on tconst identifier:

In [ ]:
data = pd.merge(basics, ratings, on ="tconst")
data.head()

tconst titleType  ... averageRating numVotes
0  tt0000001     short  ...           5.7     1858
1  tt0000002     short  ...           6.0      243
2  tt0000003     short  ...           6.5     1630
3  tt0000004     short  ...           6.0      158
4  tt0000005     short  ...           6.2     2456

[5 rows x 11 columns]

In [ ]:
data.shape[0]

1214631

Filtering TV Series data from other categories of titleType. 

In [ ]:
titleType_arr = np.array(pd.Categorical(data['titleType']).categories)

In [ ]:
print(titleType_arr)

['movie' 'short' 'tvEpisode' 'tvMiniSeries' 'tvMovie' 'tvSeries' 'tvShort'
 'tvSpecial' 'video' 'videoGame']


In [ ]:
series_data = data[data['titleType'] == 'tvSeries']

In [ ]:
series_data.shape[0]

80140

In [ ]:
series_data.head()

tconst titleType  ... averageRating numVotes
9933   tt0025509  tvSeries  ...           8.3     1759
18171  tt0035803  tvSeries  ...           7.9       51
20925  tt0039120  tvSeries  ...           3.0       14
20926  tt0039123  tvSeries  ...           8.5      185
20927  tt0039125  tvSeries  ...           5.4       17

[5 rows x 11 columns]

Removing Unnecessary Columns:

In [ ]:
series_data = series_data.drop(["titleType", "originalTitle", "isAdult", "startYear", "endYear", "runtimeMinutes", "genres", "averageRating"], axis=1)

Sorting the Series dataframe based on descending order with numVotes column to help Raghav sort the series in order based on max number of ratings and also displaying the series with max number of votes.

In [ ]:
series_data.sort_values(by=['numVotes'], ascending=False)

tconst                 primaryTitle  numVotes
432997    tt0944947              Game of Thrones   1950970
422789    tt0903747                 Breaking Bad   1666749
967246    tt4574334              Stranger Things    962273
80435     tt0108778                      Friends    934697
678143    tt1520211             The Walking Dead    928896
...             ...                          ...       ...
666377   tt14861366                  Studio 3:16         5
752943    tt1853136                   Futch Tube         5
579634   tt12747896   Mere Husband Kee Dulhaniya         5
1024038   tt5556834               Adventure Room         5
510653   tt11226238  Fox Comedy Club: Rive Rouge         5

[80140 rows x 3 columns]

In [ ]:
series_data.sort_values(by=['numVotes'], ascending=False).head(1)

tconst     primaryTitle  numVotes
432997  tt0944947  Game of Thrones   1950970

The **TV series** that received the most number of votes is **Game of Thrones** which has **1950970** votes.



**Problem 2:**

**Amber was addicted to the word game Wordle (https://www.nytimes.com/games/wordle/index.html) which has caught the (English speaking) world by storm, with the game recently being sold to the New York Times for millions of dollars. Can you utilise your knowledge in linguistics/computer science/NLP/information theory to solve this riddle more efficiently? What strategies could you use that are inspired by these fields? You need not produce any code for this - please just describe your strategy in text.**

Solving Wordle in an Efficient way:
I found many interesting ways to solve wordle. But one of the most efficient one is listed below in steps.

1. **Get the Dataset:** There is an English Word Frequency dataset in Kaggle,https://www.kaggle.com/rtatman/english-word-frequency?select=unigram_freq.csv, contains the counts of the 333,333 most commonly-used single words on the English language web, as derived from the Google Web Trillion Word Corpus. As wordle contains 5 letter words, I am restricting the dataset to the words with wordcount = 5.
2. For every round of successful guessing, we learn:
*   Letters that are in the solution
*   Letters that are not in the solution
*   Places where certain letters do appear
*  Places where certain letters do not appear
*  Number of times a letter appears in the solution

The optimal way to guess for the first time is to use the word with most common letters like 'STORY', 'RATES', 'ADIEU', 'RAISE'. 
If you pick a word with the most common letters and one or more of those common letters is not in the solution, then you have identified a large number of impossible solutions. This greatly reduces the number of possible solutions, which makes picking the correct solution in the next round more likely.
So, the next step is to focus on letter frequency.
3. **Letter frequency:** Get the frequency of each letter by joining all the words into a single very long string and using count function to see how often each letter occurs. Now we can see which are the most frequent letters and which are the least frequent ones.
4.**Position Based Score:** The next step is to get a score for each letter in each of the five positions. Now we have a score for each letter in each position. 
5.**Word Score Calculation:** Now for each word, we can calculate a score by adding each letter score based on its position. For example, for "RAISE',we can look up R and First position score + A and Second position score + and so on.
This score helps us rank the words based on the letter's frequency.
6.**Applying our findings and conditions to the Wordle:** This is the step where we actually enter the word with the maximum score from our last step to the Wordle. Now based on our result, we add more conditions to our WordScores object from the last step. i.e.if the word that we entered was 'SORES' and out of which 'O' was green and rest were grey, that means that 
**Finding 1**: 'O' is in the second and right position and
**Finding 2**: The letters "S,R,E" does not exist in out solution. 
After adding our new findings to our conditions and looking at our highest score, we guess and enter the next word and repeat till we get all the letters right.

Although this method needs some preprocessing before actually applying the words to the wordle, it can give better results than starting from random guesses.






